### PR TITLE
Add STEGO & Mask R-CNN API models

### PR DESCRIPTION
**Pull request by lichtefeld**
* Adds a server to process STEGO & Mask R-CNN models on various inputs
* Adds a script to process a curriculum through the API

### CODE REVIEW

**Review by spigo900 — COMMENTED** [review]
One comment for now: PyCharm seems to have gone out of control with a rename refactor of `model->stego_model`. (Frustrating when that happens.) I'm planning to review the new scripts sometime tomorrow.

**Comment by lichtefeld** [comment]
🙃 Ok, I'll aim to fix PyCharm's greedy renaming process before you review tomorrow.

File: segmentation_processing/mask_rcnn/utils.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    thresholded_preds_inidices = [scores.index(i) for i in scores if i > threshold]
+    thresholded_preds_count = len(thresholded_preds_inidices)
+
+    # threshold filter masks
+    masks = (outputs[0]["masks"] > 0.5).squeeze().detach().cpu().numpy()
+    masks = masks[:thresholded_preds_count]
+
+    # retrieve bounding boxes in form (x1, y1), (x2, y2)
+    boxes = [
+        [(int(i[0]), int(i[1])), (int(i[2]), int(i[3]))] for i in outputs[0]["boxes"].detach().cpu()
```
Comment (spigo900) [inline segmentation_processing/mask_rcnn/utils.py:119]: Nit: Would like to rename `i` in this context to `box_flat` say.

File: segmentation_processing/mask_rcnn/utils.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    "scissors",
+    "teddy bear",
+    "hair drier",
+    "toothbrush",
+]
+
+COLORS = np.random.uniform(0, 255, size=(len(INSTANCE_CATEGORIES), 3))
+
+
+def inference_res(image: Image, model: Any, threshold: float) -> Tuple[Any, Any, Any]:
```
Comment (spigo900) [inline segmentation_processing/mask_rcnn/utils.py:104]: Nit: This takes `img` as a PyTorch tensor AFAICT:
```suggestion
def inference_res(image: torch.Tensor, model: Any, threshold: float) -> Tuple[Any, Any, Any]:
```

**Review by spigo900 — CHANGES_REQUESTED** [review]
Looks good to me overall, just a few nitpicks. I didn't review the full ~2500 line details of Blake's STEGO implementation, but I did look a little at the inference code since that's simple enough to check -- that's again just a few nitpicks.

File: adam/curriculum/generate_segmentation_results.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+        # combine masks
+        segmentation_map = np.stack([r_map, g_map, b_map], axis=2)
+        img = np.array(img)
+        img = cv2.cvtColor(img, cv2.COLOR_RGB2BGR)
+
+        # instance mask
+        cv2.addWeighted(img, alpha, segmentation_map, beta, gamma, img)
+
+    return img
```
[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam/curriculum/generate_segmentation_results.py:51]: Ah, I meant to ask if you thought the stroke extraction code would better handle the output from this post-processing method or the RCNN post-processing? Either way I'll probably leave both in here (with a comment explaining the unused function) in case we discover a need to modify the output of this for stroke extraction. I'd like to easily have the reference available.

File: segmentation_processing/start_server.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+#SBATCH --output=./logs/%x-%j.out    # %x-%j means JOB_NAME-JOB_ID.
+
+set -euo pipefail
+
+# Indefinite timeout used to keep the stego_model in memory for a long time
+
+umask 0002  # Allow files to be created as group-writable
+../venv/bin/gunicorn \
+  --config ./gunicorn.conf.py \
+  --access-logfile ./server.log \
```
[the remark this answers is not in the document]
Comment (lichtefeld) [reply segmentation_processing/start_server.sh:19]: Right, there's also a 'small' problem with this script which is we don't use a `venv` rather that invocation to gunicorn needs to be updated to point at someone's environment on the cluster... 

Ideally we'd create a shared environment on the the `nas` for this script to use but that may be easier to manually configure than create a script for.

File: adam/curriculum/generate_segmentation_results.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+import requests
+import yaml
+from tqdm import tqdm
+
+API_ROUTE = "/instanceseg"
+DESCRIPTION_FILENAME = "description.yaml"
+INFO_FILENAME = "info.yaml"
+COLOR_SEED = 42
+LABEL_COLORS = np.random.uniform(0, 255, size=(50, 3))
```
[the remark this answers is not in the document]
Comment (spigo900) [reply adam/curriculum/generate_segmentation_results.py:21]: Since we're keeping the STEGO function, a nit: This is slightly nondeterministic since there's no seeding.

File: adam/curriculum/generate_segmentation_results.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    beta = 0.6
+    alpha = 1.0
+    gamma = 0.0
+    for mask in masks:
+        r_map = np.zeros_like(mask).astype(np.uint8)
+        g_map = np.zeros_like(mask).astype(np.uint8)
+        b_map = np.zeros_like(mask).astype(np.uint8)
+
+        # mask color assignment
+        color = LABEL_COLORS[random.randrange(0, len(LABEL_COLORS))]
```
Comment (spigo900) [inline adam/curriculum/generate_segmentation_results.py:40]: Is there a reason to use `randrange()` over deterministically indexing the label colors by (for example) the mask index? If so this should use a seed for picking the color.

**Review by spigo900 — CHANGES_REQUESTED** [review]

File: adam/curriculum/generate_segmentation_results.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+        # Copy over the depth.png, description.yaml, feature.yaml, rgb.png files if not inplace
+        if not args.inplace:
+            files_to_copy = [sit_dir / DESCRIPTION_FILENAME]
+            files_to_copy.extend(rgb_files_to_process)
+            files_to_copy.extend(sorted(sit_dir.glob("depth_*.png")))
+            files_to_copy.extend(sorted(sit_dir.glob("feature*.yaml")))
+
+            for file in files_to_copy:
+                shutil.copy2(file, sit_out_dir / file.name)
```
[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam/curriculum/generate_segmentation_results.py:163]: I'll remove it. I used it to easily compare the RGB to segmentation images.

File: adam/curriculum/generate_segmentation_results.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    )
+    parser.add_argument(
+        "--api",
+        type=str,
+        default="http://saga03.isi.edu:5001",
+        help="The server hostname & port to post API requests to.",
+    )
+    args = parser.parse_args()
+
+    base_curriculum = Path(args.base_curriculum)
```
[the remark this answers is not in the document]
Comment (spigo900) [reply adam/curriculum/generate_segmentation_results.py:126]: Ah, I see. :(

File: adam/curriculum/generate_segmentation_results.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (spigo900) [reply adam/curriculum/generate_segmentation_results.py:163]: That makes sense.

File: adam/curriculum/generate_segmentation_results.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (spigo900) [reply adam/curriculum/generate_segmentation_results.py:126]: Probably worth a comment to that effect.

File: segmentation_processing/start_server.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (spigo900) [reply segmentation_processing/start_server.sh:19]: Ah, I totally glossed over the `venv`. As far as manually configuring -- you mean activate this before running the script?

Would it be straightforward to run this in the same env as other preprocessing code? I guess that's a question for me to answer as well 🙂 -- if I just install Gunicorn/necessary other things, does it work.

[the remark this answers is not in the document]
Comment (lichtefeld) [reply segmentation_processing/start_server.sh:19]: The requirements file in this directory has what I think needs to be installed for the server to run. It may work in another environment I just only tested in a new one for simplicity. 

On the cluster my submit script points to the `gunicorn` bin file in my conda environment. We may be able to run this by just activating the environment `conda activate` and running `gunicorn` on the command line, I didn't test that. I just modified the scripts LESTAT uses.

File: adam/curriculum/generate_segmentation_results.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    if not base_curriculum.is_dir():
+        raise OSError(f"{base_curriculum} is not a directory.")
+
+    if args.inplace and args.copy_to:
+        raise RuntimeError(
+            "Unable to both copy curriculum to a new directory and place segmentation results inplace."
+        )
+
+    output_cur_path = base_curriculum if args.inplace else Path(args.copy_to)
```
[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam/curriculum/generate_segmentation_results.py:135]: Taking your suggestion to remove the `--inplace` option

File: adam/curriculum/generate_segmentation_results.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
+import argparse
+import base64
+import json
+import logging
+import shutil
```
Comment (spigo900) [inline adam/curriculum/generate_segmentation_results.py:5]: Can get rid of this import now?
```suggestion
```

**Review by spigo900 — COMMENTED** [review]
Looks good to me. One minor thing that precommit will probably pick up.

Other than that, we should decide what we're doing about gunicorn. As-is the server startup script needs a local virtualenv to run. Do we want to keep that, set up a global env, or switch to a conda env? Something else?

**Comment by lichtefeld** [comment]
It may be best to setup a shared-env on the cluster which we all have access to so that submitting the job for the server can be done by any of us. We'll have to manage by hand making sure we don't start overlapping servers or provide a way to pass in a port number but that should be doable and I'm unsure a more sophisticated automation solution is needed for this project.

**Comment by spigo900** [comment]
@lichtefeld Sounds good to me. Will you set up the shared-env? And would the shared-env go in the script or in the README? I'm thinking the script makes sense since it's shared.

**Comment by lichtefeld** [comment]
@spigo900 I can set up the shared env if that's preferred. I'd point this submission script at environment on the cluster so there's no modifications needed. I'll include instructions on configuring a local shared environment in the readme for completeness.

**Review by spigo900 — APPROVED** [review]
Looks good to me. I tested the setup script on the cluster and it works for me.

File: segmentation_processing/mask_rcnn/utils.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
+from typing import Any, Tuple
+
+import numpy as np
+import torch
+
+INSTANCE_CATEGORIES = [
```
Comment (boyleconnor) [inline segmentation_processing/mask_rcnn/utils.py:6]: Is there a reason why this is defined here and in `instance_names.py`?

**Review by boyleconnor — CHANGES_REQUESTED** [review]
I did not review super thoroughly, I mostly just looked at the README.

File: segmentation_processing/mask_rcnn/utils.py (hunk 11 above)
Comment (lichtefeld) [reply segmentation_processing/mask_rcnn/utils.py:6]: The Mask RCNN's instances names could differ from the STEGO ones. All I did was relocate model code within its given directory I didn't try to unify between the two.

**Comment by lichtefeld** [comment]
@boyleconnor I've reorganized the Readme some as well as merged in suggestions of yours.

File: segmentation_processing/README.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+## Cluster Configuration
+
+When creating the application's `venv`, this Python can be easily used with the following command run in `segmentation_procssing`:
+
+```bash
+PATH="/nas/gaia/adam/shared/local_python/bin/:$PATH" make install
+```
+
+### Creating and upgrading a local Python instance
```
[the remark this answers is not in the document]
Comment (boyleconnor) [reply segmentation_processing/README.md:24]: I don't believe this comment was addressed

File: segmentation_processing/README.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+The following commands were used to create a recent local Python:
+
+```bash
+pyenv install miniconda3-3.8-4.12.0
+pyenv local miniconda3-3.8-4.12.0
+conda create --channel conda-forge --prefix ./local_python python=3.8.13
+pyenv local --unset
+```
+
+When updating to a new minor version of Python, the original `venv` needs to be deleted first or else no new `venv` will be created. The `venv` cannot be upgraded easily in-place because most non-pure Python packages must be reinstalled to work on a different minor version.
```
Comment (boyleconnor) [inline segmentation_processing/README.md:37]: I think everything from this line down is "how to keep the Python env up-to-date" and therefore should have it's own header. Correct?

File: segmentation_processing/README.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+With `segmentation_processing` as the working directory, run the following command, which will create a virtual environment (`venv`) with the necessary dependencies in the working directory, as well as download the STEGO model.
+
+```bash
+PATH="/nas/gaia/adam/shared/local_python/bin/:$PATH" make install
+```
+
+### Creating and upgrading a local Python instance
+
+The easiest way we've found is to create a local Python instance is with Conda, which also has the advantage of using a pre-compiled `python` executable that has no linking issues.
```
Comment (boyleconnor) [inline segmentation_processing/README.md:26]: Is this an alternative to the steps above? Does it need to be done beforehand? Please clarify this

**Review by boyleconnor — CHANGES_REQUESTED** [review]
Again, I mostly just looked at the README. I think it would be difficult to follow these installation/update instructions.

File: segmentation_processing/README.md (hunk 14 above)
Comment (lichtefeld) [reply segmentation_processing/README.md:26]: This is the instructions to setup a local python instance on the cluster which isn't linked against any system library. 

I followed the Lestat documentation trend of putting the less-frequently followed setup steps later in the readme. This is intended to mostly be a documentation of how we configured our server, any individual deployment by someone in the future has other details to work out which I can't document.

File: segmentation_processing/README.md (hunk 13 above)
Comment (lichtefeld) [reply segmentation_processing/README.md:37]: This is how to update the base python version within the `venv` not just the requirements of the venv. It's specific to the creation process this header already has.

File: segmentation_processing/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (lichtefeld) [reply segmentation_processing/README.md:24]: I think I may have addressed this comment in my replies below? Please let me know if you're still unclear

File: segmentation_processing/README.md (hunk 14 above)
Comment (boyleconnor) [reply segmentation_processing/README.md:26]: @lichtefeld could you add flow sentences under each header saying which steps are less-frequently followed and why? I think it would also be helpful to put a statement to this effect:

> any individual deployment by someone in the future has other details to work out which I can't document.

at the top of the README.

**Review by boyleconnor — APPROVED** [review]

**Comment by boyleconnor** [comment]
@lichtefeld I approved so you don't have to come back to get my feedback again but I recommend some changes in a comment
